File: solutions/elixir/resistor-color-trio/1/lib/resistor_color_trio.ex
# For Shree DR.MDD

defmodule ResistorColorTrio do
  @color_values %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @giga_unit 1_000_000_000
  @mega_unit 1_000_000
  @kilo_unit 1_000

  @spec label(colors :: [atom]) :: {integer, :ohms | :kiloohms | :megaohms | :gigaohms}
  def label([first_color | [second_color | [mult_color | _]]]) do
    d1 = @color_values[first_color]
    d2 = @color_values[second_color]
    zeroes = List.duplicate(0, @color_values[mult_color])

    total_ohms = Integer.undigits([d1 | [d2 | zeroes]])

    cond do
      total_ohms >= @giga_unit -> {div(total_ohms, @giga_unit), :gigaohms}
      total_ohms >= @mega_unit -> {div(total_ohms, @mega_unit), :megaohms}
      total_ohms >= @kilo_unit -> {div(total_ohms, @kilo_unit), :kiloohms}
      true -> {total_ohms, :ohms}
    end
  end
end

File: solutions/julia/resistor-color-trio/1/resistor-color-trio.jl
function label(colors)
    cols = ["black", "brown", "red", "orange", "yellow", "green", "blue", "violet", "grey", "white"]
    res = ""
    d = Dict(cols[a+1] => a for a in 0:9)
    val = (d[colors[1]]*10 + d[colors[2]]) * 10^d[colors[3]]
    if val > 1e9
        hold = (val/1e9)
        if isinteger(hold)
            hold = Int(hold)
        end            
        res = "$hold gigaohms"
    elseif val > 1e6
        hold = (val/1e6)
        if isinteger(hold)
            hold = Int(hold)
        end            
        res = "$hold megaohms"
    elseif val > 1e3
        hold = (val/1e3)
        if isinteger(hold)
            hold = Int(hold)
        end            
        res = "$hold kiloohms"
    else
        res = "$(val) ohms"
    end
    res
end